paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+            else:
+                filtered_lines.append(line)
+
+        self.template_lines = filtered_lines
+
+
+    def write_input(self, include_saves=True):
+        """
+        Write a tleap input file based on template_lines and other things we have set.
+
+        """
+
+        file_path = self.output_path + self.output_prefix + '.tleap.in'
+        with open(file_path, 'w') as f:
+            for line in self.template_lines:
+                f.write(line+"\n")
+    
+            if self.pbc_type == 'cubic':
+                f.write("solvatebox {} {} {} iso\n".format(self.unit, self.water_box, self.buffer_value))
+            elif self.pbc_type == 'rectangular':
+                f.write("solvatebox {} {} {{10.0 10.0 {}}}\n".format(self.unit, self.water_box, self.buffer_value))
+            elif self.pbc_type == 'octahedral':
+                f.write("solvateoct {} {} {} iso\n".format(self.unit, self.water_box, self.buffer_value))
+            elif self.pbc_type is None:
+                f.write("# Skipping solvation ...\n")
+            else:
+                raise Exception(
+                    "Incorrect pbctype value provided: " + str(self.pbc_type) \
+                    + ". Only `cubic`, `rectangular`, `octahedral`, and None are valid")
+            if self.neutralize:
+                f.write("addionsrand {} {} 0\n".format(self.unit, self.counter_cation))
+                f.write("addionsrand {} {} 0\n".format(self.unit, self.counter_anion))
+            # Additional ions should be specified as a list, with residue name and number of ions in pairs, like ['NA',
+            # 5] for five additional sodium ions. By this point, if the user specified a molality or molarity,
+            # it should already have been converted into a number.
+            if self.add_ion_residues:
+                for residue, amount in zip(self.add_ion_residues[0::2],
+                                           self.add_ion_residues[1::2]):
+                    f.write("addionsrand {} {} {}\n".format(self.unit, residue, amount))
+            if self.waters_to_remove:
+                for water_number in self.waters_to_remove:
+                    f.write("remove {} {}.{}\n".format(self.unit, self.unit, water_number))
+    
+            # Note, the execution of tleap is assumed to take place in the
+            # same directory as all the associated input files, so we won't
+            # put directory paths on the saveamberparm or savepdb commands.
+            if self.output_prefix and include_saves:
+                f.write("savepdb {} {}.pdb\n".format(self.unit, self.output_prefix))
+                f.write("saveamberparm {} {}.prmtop {}.rst7\n".format(self.unit, self.output_prefix,
+                                                                      self.output_prefix))
+            else:
+                pass
+            f.write("desc {}\n".format(self.unit))
+            f.write("quit\n")
+
+    def run(self):
```

Comment:
Update docstring with return type and whatever it is.

Reply:
Done